subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy

```diff
@@ -62,12 +62,36 @@ class JavaReflectionUtilTest extends Specification {
 
     def "write property"() {
         when:
-        writeableProperty(JavaTestSubject, "myProperty").setValue(myProperties, "otherValue")
+        writeableProperty(JavaTestSubject, "myProperty", String.class).setValue(myProperties, "otherValue")
 
         then:
         readableProperty(JavaTestSubject, String, "myProperty").getValue(myProperties) == "otherValue"
     }
 
+    def "picks the generic object setter if the typed setter does not match the value type"() {
+        when:
+        def property = writeableProperty(JavaTestSubject, "myProperty", File.class)
+
+        then:
+        property.type == Object.class
+    }
+
+    def "picks the generic iterable setter if the typed setter does not match the value type"() {
+        when:
+        def property = writeableProperty(JavaTestSubject, "multiValue", List.class)
+
+        then:
+        property.type == Iterable.class
+    }
+
+    def "can handle null as property type"() {
```

Comment:
Unfortunately, it is not deterministic which setter gets called. All are equally correct. Depends again of the order the methods are returned by `Class.getMethods()` - see implementation of `getMatchingAccessibleMethod()` in commons lang. I think we need to accept this.

I added a test for the primitive type case. (It will not find a suitable setter as null cannot be assigned.)